File: tests/algorithms/transpiler/FS/cellular_automata/wa_tor.fs
// Generated 2025-08-06 22:46 +0700

exception Break
exception Continue

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let WIDTH: int = 10
let HEIGHT: int = 10
let PREY_INITIAL_COUNT: int = 20
let PREY_REPRODUCTION_TIME: int = 5
let PREDATOR_INITIAL_COUNT: int = 5
let PREDATOR_REPRODUCTION_TIME: int = 20
let PREDATOR_INITIAL_ENERGY: int = 15
let PREDATOR_FOOD_VALUE: int = 5
let TYPE_PREY: int = 0
let TYPE_PREDATOR: int = 1
let mutable seed: int = 123456789
let rec rand () =
    let mutable __ret : int = Unchecked.defaultof<int>
    try
        seed <- int ((((int64 ((seed * 1103515245) + 12345)) % 2147483648L + 2147483648L) % 2147483648L))
        __ret <- seed
        raise Return
        __ret
    with
        | Return -> __ret
and rand_range (max: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable max = max
    try
        __ret <- (((rand()) % max + max) % max)
        raise Return
        __ret
    with
        | Return -> __ret
and shuffle (list_int: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable list_int = list_int
    try
        let mutable i: int = (Seq.length (list_int)) - 1
        while i > 0 do
            let mutable j: int = rand_range (i + 1)
            let tmp: int = _idx list_int (i)
            list_int <- _arrset list_int i (_idx list_int (j))
            list_int <- _arrset list_int j (tmp)
            i <- i - 1
        __ret <- list_int
        raise Return
        __ret
    with
        | Return -> __ret
and create_board () =
    let mutable __ret : int array array = Unchecked.defaultof<int array array>
    try
        let mutable board: int array array = [||]
        let mutable r: int = 0
        while r < HEIGHT do
            let mutable row: int array = [||]
            let mutable c: int = 0
            while c < WIDTH do
                row <- Array.append row [|0|]
                c <- c + 1
            board <- Array.append board [|row|]
            r <- r + 1
        __ret <- board
        raise Return
        __ret
    with
        | Return -> __ret
and create_prey (r: int) (c: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable r = r
    let mutable c = c
    try
        __ret <- unbox<int array> [|TYPE_PREY; r; c; PREY_REPRODUCTION_TIME; 0; 1|]
        raise Return
        __ret
    with
        | Return -> __ret
and create_predator (r: int) (c: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable r = r
    let mutable c = c
    try
        __ret <- unbox<int array> [|TYPE_PREDATOR; r; c; PREDATOR_REPRODUCTION_TIME; PREDATOR_INITIAL_ENERGY; 1|]
        raise Return
        __ret
    with
        | Return -> __ret
let mutable board: int array array = create_board()
let mutable entities: int array array = [||]
let rec empty_cell (r: int) (c: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable r = r
    let mutable c = c
    try
        __ret <- (_idx (_idx board (r)) (c)) = 0
        raise Return
        __ret
    with
        | Return -> __ret
and add_entity (typ: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable typ = typ
    try
        while true do
            let mutable r: int = rand_range (HEIGHT)
            let mutable c: int = rand_range (WIDTH)
            if empty_cell (r) (c) then
                if typ = TYPE_PREY then
                    board.[r].[c] <- 1
                    entities <- Array.append entities [|create_prey (r) (c)|]
                else
                    board.[r].[c] <- 2
                    entities <- Array.append entities [|create_predator (r) (c)|]
                __ret <- ()
                raise Return
        __ret
    with
        | Return -> __ret
and setup () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let mutable i: int = 0
        while i < PREY_INITIAL_COUNT do
            add_entity (TYPE_PREY)
            i <- i + 1
        i <- 0
        while i < PREDATOR_INITIAL_COUNT do
            add_entity (TYPE_PREDATOR)
            i <- i + 1
        __ret
    with
        | Return -> __ret
let dr: int array = [|-1; 0; 1; 0|]
let dc: int array = [|0; 1; 0; -1|]
let rec inside (r: int) (c: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable r = r
    let mutable c = c
    try
        __ret <- (((r >= 0) && (r < HEIGHT)) && (c >= 0)) && (c < WIDTH)
        raise Return
        __ret
    with
        | Return -> __ret
and find_prey (r: int) (c: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable r = r
    let mutable c = c
    try
        let mutable i: int = 0
        while i < (Seq.length (entities)) do
            let e: int array = _idx entities (i)
            if ((((_idx e (5)) = 1) && ((_idx e (0)) = TYPE_PREY)) && ((_idx e (1)) = r)) && ((_idx e (2)) = c) then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
and step_world () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let mutable i: int = 0
        try
            while i < (Seq.length (entities)) do
                try
                    let mutable e: int array = _idx entities (i)
                    if (_idx e (5)) = 0 then
                        i <- i + 1
                        raise Continue
                    let typ: int = _idx e (0)
                    let mutable row: int = _idx e (1)
                    let col: int = _idx e (2)
                    let repro: int = _idx e (3)
                    let energy: int = _idx e (4)
                    let mutable dirs: int array = [|0; 1; 2; 3|]
                    dirs <- shuffle (dirs)
                    let mutable moved: bool = false
                    let old_r: int = row
                    let old_c: int = col
                    if typ = TYPE_PREDATOR then
                        let mutable j: int = 0
                        let mutable ate: bool = false
                        try
                            while j < 4 do
                                try
                                    let d: int = _idx dirs (j)
                                    let nr: int = row + (_idx dr (d))
                                    let nc: int = col + (_idx dc (d))
                                    if (inside (nr) (nc)) && ((_idx (_idx board (nr)) (nc)) = 1) then
                                        let prey_index: int = find_prey (nr) (nc)
                                        if prey_index >= 0 then
                                            entities.[prey_index].[5] <- 0
                                        board.[nr].[nc] <- 2
                                        board.[row].[col] <- 0
                                        e <- _arrset e 1 (nr)
                                        e <- _arrset e 2 (nc)
                                        e <- _arrset e 4 ((energy + PREDATOR_FOOD_VALUE) - 1)
                                        moved <- true
                                        ate <- true
                                        raise Break
                                    j <- j + 1
                                with
                                | Continue -> ()
                                | Break -> raise Break
                        with
                        | Break -> ()
                        | Continue -> ()
                        if not ate then
                            j <- 0
                            try
                                while j < 4 do
                                    try
                                        let d: int = _idx dirs (j)
                                        let nr: int = row + (_idx dr (d))
                                        let nc: int = col + (_idx dc (d))
                                        if (inside (nr) (nc)) && ((_idx (_idx board (nr)) (nc)) = 0) then
                                            board.[nr].[nc] <- 2
                                            board.[row].[col] <- 0
                                            e <- _arrset e 1 (nr)
                                            e <- _arrset e 2 (nc)
                                            moved <- true
                                            raise Break
                                        j <- j + 1
                                    with
                                    | Continue -> ()
                                    | Break -> raise Break
                            with
                            | Break -> ()
                            | Continue -> ()
                            e <- _arrset e 4 (energy - 1)
                        if (_idx e (4)) <= 0 then
                            e <- _arrset e 5 (0)
                            board.[_idx e (1)].[_idx e (2)] <- 0
                    else
                        let mutable j: int = 0
                        try
                            while j < 4 do
                                try
                                    let d: int = _idx dirs (j)
                                    let nr: int = row + (_idx dr (d))
                                    let nc: int = col + (_idx dc (d))
                                    if (inside (nr) (nc)) && ((_idx (_idx board (nr)) (nc)) = 0) then
                                        board.[nr].[nc] <- 1
                                        board.[row].[col] <- 0
                                        e <- _arrset e 1 (nr)
                                        e <- _arrset e 2 (nc)
                                        moved <- true
                                        raise Break
                                    j <- j + 1
                                with
                                | Continue -> ()
                                | Break -> raise Break
                        with
                        | Break -> ()
                        | Continue -> ()
                    if (_idx e (5)) = 1 then
                        if moved && (repro <= 0) then
                            if typ = TYPE_PREY then
                                board.[old_r].[old_c] <- 1
                                entities <- Array.append entities [|create_prey (old_r) (old_c)|]
                                e <- _arrset e 3 (PREY_REPRODUCTION_TIME)
                            else
                                board.[old_r].[old_c] <- 2
                                entities <- Array.append entities [|create_predator (old_r) (old_c)|]
                                e <- _arrset e 3 (PREDATOR_REPRODUCTION_TIME)
                        else
                            e <- _arrset e 3 (repro - 1)
                    i <- i + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        let mutable alive: int array array = [||]
        let mutable k: int = 0
        while k < (Seq.length (entities)) do
            let e2: int array = _idx entities (k)
            if (_idx e2 (5)) = 1 then
                alive <- Array.append alive [|e2|]
            k <- k + 1
        entities <- alive
        __ret
    with
        | Return -> __ret
and count_entities (typ: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable typ = typ
    try
        let mutable cnt: int = 0
        let mutable i: int = 0
        while i < (Seq.length (entities)) do
            if ((_idx (_idx entities (i)) (0)) = typ) && ((_idx (_idx entities (i)) (5)) = 1) then
                cnt <- cnt + 1
            i <- i + 1
        __ret <- cnt
        raise Return
        __ret
    with
        | Return -> __ret
setup()
let mutable t: int = 0
while t < 10 do
    step_world()
    t <- t + 1
printfn "%s" ("Prey: " + (_str (count_entities (TYPE_PREY))))
printfn "%s" ("Predators: " + (_str (count_entities (TYPE_PREDATOR))))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/Kotlin/cellular_automata/wa_tor.kt
import java.math.BigInteger

fun <T> _listSet(lst: MutableList<T>, idx: Int, v: T) { while (lst.size <= idx) lst.add(v); lst[idx] = v }

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

var WIDTH: Int = 10
var HEIGHT: Int = 10
var PREY_INITIAL_COUNT: Int = 20
var PREY_REPRODUCTION_TIME: Int = 5
var PREDATOR_INITIAL_COUNT: Int = 5
var PREDATOR_REPRODUCTION_TIME: Int = 20
var PREDATOR_INITIAL_ENERGY: Int = 15
var PREDATOR_FOOD_VALUE: Int = 5
var TYPE_PREY: Int = 0
var TYPE_PREDATOR: Int = 1
var seed: Int = 123456789
var board: MutableList<MutableList<Int>> = create_board()
var entities: MutableList<MutableList<Int>> = mutableListOf<MutableList<Int>>()
var dr: MutableList<Int> = mutableListOf(0 - 1, 0, 1, 0)
var dc: MutableList<Int> = mutableListOf(0, 1, 0, 0 - 1)
fun rand(): Int {
    seed = ((Math.floorMod((((seed * 1103515245) + 12345).toLong()), 2147483648L)).toInt())
    return seed
}

fun rand_range(max: Int): Int {
    return Math.floorMod(rand(), max)
}

fun shuffle(list_int: MutableList<Int>): MutableList<Int> {
    var i: BigInteger = ((list_int.size - 1).toBigInteger())
    while (i.compareTo((0).toBigInteger()) > 0) {
        var j: Int = rand_range(((i.add((1).toBigInteger())).toInt()))
        var tmp: Int = list_int[(i).toInt()]!!
        _listSet(list_int, (i).toInt(), list_int[j]!!)
        _listSet(list_int, j, tmp)
        i = i.subtract((1).toBigInteger())
    }
    return list_int
}

fun create_board(): MutableList<MutableList<Int>> {
    var board: MutableList<MutableList<Int>> = mutableListOf<MutableList<Int>>()
    var r: Int = 0
    while (r < HEIGHT) {
        var row: MutableList<Int> = mutableListOf<Int>()
        var c: Int = 0
        while (c < WIDTH) {
            row = run { val _tmp = row.toMutableList(); _tmp.add(0); _tmp }
            c = c + 1
        }
        board = run { val _tmp = board.toMutableList(); _tmp.add(row); _tmp }
        r = r + 1
    }
    return board
}

fun create_prey(r: Int, c: Int): MutableList<Int> {
    return mutableListOf(TYPE_PREY, r, c, PREY_REPRODUCTION_TIME, 0, 1)
}

fun create_predator(r: Int, c: Int): MutableList<Int> {
    return mutableListOf(TYPE_PREDATOR, r, c, PREDATOR_REPRODUCTION_TIME, PREDATOR_INITIAL_ENERGY, 1)
}

fun empty_cell(r: Int, c: Int): Boolean {
    return (((board[r]!!) as MutableList<Int>))[c]!! == 0
}

fun add_entity(typ: Int): Unit {
    while (true) {
        var r: Int = rand_range(HEIGHT)
        var c: Int = rand_range(WIDTH)
        if (((empty_cell(r, c)) as Boolean)) {
            if (typ == TYPE_PREY) {
                _listSet(board[r]!!, c, 1)
                entities = run { val _tmp = entities.toMutableList(); _tmp.add(create_prey(r, c)); _tmp }
            } else {
                _listSet(board[r]!!, c, 2)
                entities = run { val _tmp = entities.toMutableList(); _tmp.add(create_predator(r, c)); _tmp }
            }
            return
        }
    }
}

fun setup(): Unit {
    var i: Int = 0
    while (i < PREY_INITIAL_COUNT) {
        add_entity(TYPE_PREY)
        i = i + 1
    }
    i = 0
    while (i < PREDATOR_INITIAL_COUNT) {
        add_entity(TYPE_PREDATOR)
        i = i + 1
    }
}

fun inside(r: Int, c: Int): Boolean {
    return (((((((r >= 0) && (r < HEIGHT) as Boolean)) && (c >= 0) as Boolean)) && (c < WIDTH)) as Boolean)
}

fun find_prey(r: Int, c: Int): Int {
    var i: Int = 0
    while (i < entities.size) {
        var e: MutableList<Int> = entities[i]!!
        if ((((((e[5]!! == 1) && (e[0]!! == TYPE_PREY) as Boolean)) && (e[1]!! == r) as Boolean)) && (e[2]!! == c)) {
            return i
        }
        i = i + 1
    }
    return 0 - 1
}

fun step_world(): Unit {
    var i: Int = 0
    while (i < entities.size) {
        var e: MutableList<Int> = entities[i]!!
        if (e[5]!! == 0) {
            i = i + 1
            continue
        }
        var typ: Int = e[0]!!
        var row: Int = e[1]!!
        var col: Int = e[2]!!
        var repro: Int = e[3]!!
        var energy: Int = e[4]!!
        var dirs: MutableList<Int> = mutableListOf(0, 1, 2, 3)
        dirs = shuffle(dirs)
        var moved: Boolean = false
        var old_r: Int = row
        var old_c: Int = col
        if (typ == TYPE_PREDATOR) {
            var j: Int = 0
            var ate: Boolean = false
            while (j < 4) {
                var d: Int = dirs[j]!!
                var nr: Int = row + dr[d]!!
                var nc: Int = col + dc[d]!!
                if (inside(nr, nc) && ((((board[nr]!!) as MutableList<Int>))[nc]!! == 1)) {
                    var prey_index: Int = find_prey(nr, nc)
                    if (prey_index >= 0) {
                        _listSet(entities[prey_index]!!, 5, 0)
                    }
                    _listSet(board[nr]!!, nc, 2)
                    _listSet(board[row]!!, col, 0)
                    _listSet(e, 1, nr)
                    _listSet(e, 2, nc)
                    _listSet(e, 4, (energy + PREDATOR_FOOD_VALUE) - 1)
                    moved = true
                    ate = true
                    break
                }
                j = j + 1
            }
            if (!ate) {
                j = 0
                while (j < 4) {
                    var d: Int = dirs[j]!!
                    var nr: Int = row + dr[d]!!
                    var nc: Int = col + dc[d]!!
                    if (inside(nr, nc) && ((((board[nr]!!) as MutableList<Int>))[nc]!! == 0)) {
                        _listSet(board[nr]!!, nc, 2)
                        _listSet(board[row]!!, col, 0)
                        _listSet(e, 1, nr)
                        _listSet(e, 2, nc)
                        moved = true
                        break
                    }
                    j = j + 1
                }
                _listSet(e, 4, energy - 1)
            }
            if (e[4]!! <= 0) {
                _listSet(e, 5, 0)
                _listSet(board[e[1]!!]!!, e[2]!!, 0)
            }
        } else {
            var j: Int = 0
            while (j < 4) {
                var d: Int = dirs[j]!!
                var nr: Int = row + dr[d]!!
                var nc: Int = col + dc[d]!!
                if (inside(nr, nc) && ((((board[nr]!!) as MutableList<Int>))[nc]!! == 0)) {
                    _listSet(board[nr]!!, nc, 1)
                    _listSet(board[row]!!, col, 0)
                    _listSet(e, 1, nr)
                    _listSet(e, 2, nc)
                    moved = true
                    break
                }
                j = j + 1
            }
        }
        if (e[5]!! == 1) {
            if (moved && (repro <= 0)) {
                if (typ == TYPE_PREY) {
                    _listSet(board[old_r]!!, old_c, 1)
                    entities = run { val _tmp = entities.toMutableList(); _tmp.add(create_prey(old_r, old_c)); _tmp }
                    _listSet(e, 3, PREY_REPRODUCTION_TIME)
                } else {
                    _listSet(board[old_r]!!, old_c, 2)
                    entities = run { val _tmp = entities.toMutableList(); _tmp.add(create_predator(old_r, old_c)); _tmp }
                    _listSet(e, 3, PREDATOR_REPRODUCTION_TIME)
                }
            } else {
                _listSet(e, 3, repro - 1)
            }
        }
        i = i + 1
    }
    var alive: MutableList<MutableList<Int>> = mutableListOf<MutableList<Int>>()
    var k: Int = 0
    while (k < entities.size) {
        var e2: MutableList<Int> = entities[k]!!
        if (e2[5]!! == 1) {
            alive = run { val _tmp = alive.toMutableList(); _tmp.add(e2); _tmp }
        }
        k = k + 1
    }
    entities = alive
}

fun count_entities(typ: Int): Int {
    var cnt: Int = 0
    var i: Int = 0
    while (i < entities.size) {
        if (((((entities[i]!!) as MutableList<Int>))[0]!! == typ) && ((((entities[i]!!) as MutableList<Int>))[5]!! == 1)) {
            cnt = cnt + 1
        }
        i = i + 1
    }
    return cnt
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        setup()
        var t: Int = 0
        while (t < 10) {
            step_world()
            t = t + 1
        }
        println("Prey: " + count_entities(TYPE_PREY).toString())
        println("Predators: " + count_entities(TYPE_PREDATOR).toString())
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
